dvc/remote/azure.py

```diff
@@ -18,11 +18,33 @@
 from dvc.config import Config
 from dvc.remote.base import RemoteBASE
 from dvc.path_info import CloudURLInfo
+from dvc.exceptions import DvcException
 
 
 logger = logging.getLogger(__name__)
 
 
+def _read_connection_string_from_config(config):
+    connection_string_from_config = config.get(
+        Config.SECTION_AZURE_CONNECTION_STRING
+    )
+    if (
+        connection_string_from_config is not None
+    ) and connection_string_from_config.startswith("$"):
+        # read custom environment variable discarding the '$' character
```

Comment:
Do we actually need $ though? Maybe I'm missing something, but  I am not able to find $ used in connection strings at all.

Reply:
Realised that that ^ didn't come up very clear 😅 I meant that I am not able to find where such trick with env vars would be used. Or is it something that you've come up with yourself?

Reply:
Ok, maybe a little background. When working with Azure, it is handy to save the storage account credentials as connection strings in a environment variables and retrieve them from client libraries (like [`azure-storage-blob`](https://pypi.org/project/azure-storage-blob/)).  I explicitly used plural form to highlight the fact that there are multiple storage accounts. Now working with dvc remotes on different storages you need a way to select the right destination of data. The solution with ENV vars would keep the connection string private while showing the name of the ENV variable (which itself encodes the destination with e.g. the storage-account name). Team members can, after setting up their `.bash_profile`, contribute to projects.

You are right, it is nothing Azure specific - maybe a helper function for other remote types would be handy...

Reply:
For other remotes it might be dangerous. What if someone would want to do something like $PART1_$PART2 ? This env evaluation in config feels like it is too powerful and doesn't really belong in the current form in the config file itself. It is also weird that it will be supported like that for 1 specific config option for 1 specific remote and if we try to support it for other options later, it has a potential to run into conflicts, where we might no longer be able to tell easily if it is an env var that we need to evaluate. But just mixing config and env doesn't feel right in this scenario.

Feels like the proper way to work with this is to use .dvc/config.local, as you've described. You will be sharing your .bash_profile anyway, so why not just share .dvc/config.local privately instead? You can use placeholders in .dvc/config like so too if you want to:
```
dvc remote modify myremote connection_string MY_PLACEHOLDER
```

Reply:
> For other remotes it might be dangerous. What if someone would want to do something like $PART1_$PART2 ? This env evaluation in config feels like it is too powerful and doesn't really belong in the current form in the config file itself. It is also weird that it will be supported like that for 1 specific config option for 1 specific remote and if we try to support it for other options later, it has a potential to run into conflicts, where we might no longer be able to tell easily if it is an env var that we need to evaluate. But just mixing config and env doesn't feel right in this scenario.
> 
> Feels like the proper way to work with this is to use .dvc/config.local, as you've described. You will be sharing your .bash_profile anyway, so why not just share .dvc/config.local privately instead? You can use placeholders in .dvc/config like so too if you want to:
> 
> ```
> dvc remote modify myremote connection_string MY_PLACEHOLDER
> ```

I get your first point about the power of such parsing functionality - agree on that completely.

On the other hand (your second point) there exists a specific solution with azure where we read a connection string from ENV with `AZURE_STORAGE_CONNECTION_STRING` - working with one storage account will be fine. Having different storage accounts with different levels of security/back up strategy will be hard (since we distribute write/delete/list connection strings).

Why I do not want to share .dvc/config.local is because you have to do that for each project - plus you have to track who got which token for which project instead of having a list of team members and tokens once.

Reply:
I agree with your points, but I'm just not sure that this is the proper way to solve your problem. With things like s3, you would probably just use profiles, right? Does azure have those?

Reply:
@kopytjuk Or maybe file-based auth?

Reply:
No Azure client libraries (at least the official one in Python) supports authentication via SAS tokens or primary key of the storage account.

Reply:
@kopytjuk maybe something like 
```
client = get_client_from_auth_file(StorageManagementClient)
storage_keys = storage_client.storage_accounts.list_keys(resource_group_name, storage_account_name)
storage_keys = {v.key_name: v.value for v in storage_keys.keys}

storage_client = CloudStorageAccount(storage_account_name, storage_keys['key1'])
blob_service = storage_client.create_block_blob_service()
```
from https://github.com/Azure/azure-sdk-for-python/issues/1310#issuecomment-316551927 ?